### PR TITLE
feat(provider/google-vertex): add Chirp 3 HD voices (Cloud Text-to-Speech) speech model

### PR DESCRIPTION
**Pull request by 31Carlton7**
## Motivation

Google's Chirp 3: HD voices are served by the **Google Cloud Text-to-Speech API** (`text:synthesize`), not the Vertex Gemini `generateContent` endpoint that the existing `@ai-sdk/google-vertex` speech path uses — so they need a dedicated model implementation.

## Design

- New `GoogleVertexCloudTTSSpeechModel` (`packages/google-vertex/src/google-vertex-cloud-tts-speech-model.ts`) implementing `SpeechModelV4` against `POST https://texttospeech.googleapis.com/v1/text:synthesize` (single non-regional host), reusing the provider's Google Cloud auth headers / fetch — the same config plumbing as the Vertex transcription (Speech-to-Text) model.
- **Routing**: `createSpeechModel` in `google-vertex-provider-base.ts` routes model ids starting with `chirp` (e.g. the new `chirp-3-hd` id) to the Cloud TTS model; all `gemini-*` ids keep the existing Gemini TTS path unchanged.
- **Voice composition**: Chirp 3: HD voices are named `<locale>-Chirp3-HD-<voice>`. The model composes the name from the standard call options — `language` (BCP-47 locale, default `en-US`) + `voice` (default `Kore`) → `en-US-Chirp3-HD-Kore`. A fully-qualified voice name containing `Chirp3-HD` is passed through verbatim (its locale prefix is used as `languageCode` unless `language` is set).
- **Audio**: requested as `LINEAR16`, which Cloud TTS returns as a WAV (RIFF) file — directly playable; base64 `audioContent` is decoded to bytes. `speed` maps to `audioConfig.speakingRate`. `instructions` and non-`wav` output formats produce `unsupported` warnings.
- **Usage**: Cloud TTS bills per input character. No token usage is fabricated; the raw API response is preserved on `response.body` and the request text is available via `request.body`, so callers can derive character usage from the input text.

## Out of scope (intentionally)

- `streamingSynthesize` (bidi streaming synthesis)
- Instant Custom Voice / voice cloning
- Non-Chirp Cloud TTS voice families (Standard/WaveNet/Neural2/Studio) — the routing is `chirp*`-prefix based, so these can be layered on later if wanted

## Tests

`google-vertex-cloud-tts-speech-model.test.ts` (mirrors the transcription model test style):
- endpoint URL + provider/request header + user-agent assertions
- request body shape (input/voice/audioConfig with LINEAR16)
- voice-name composition (`language` + `voice` → `en-US-Chirp3-HD-Kore`; verbatim pass-through of fully-qualified names; explicit `language` precedence)
- base64 → bytes audio decode, empty-audio fallback
- `speed` → `speakingRate` mapping
- warnings for `instructions` and unsupported output formats
- response metadata (timestamp, modelId, headers, raw body) and API error handling

Plus provider-base routing tests: `chirp-3-hd` → Cloud TTS model, `gemini-*` → existing Gemini speech model.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by ManyaS-Git — COMMENTED** [review]
## Code Review: feat(provider/google-vertex): add Chirp 3 HD voices (Cloud Text-to-Speech) speech model

Thanks for the thorough implementation! The overall structure is solid, the tests are well-written, and the voice-name composition logic is clear. I have a few suggestions below.

---

### 1. Missing Express Mode (API key) guard -- should fix

The Cloud Text-to-Speech API (`texttospeech.googleapis.com`) is a **separate Google Cloud API**, not the Vertex AI `generateContent` endpoint. Express Mode API keys only work with the Vertex AI endpoint. The transcription model (`google-vertex-provider-base.ts:335-352`) already guards against this:

```typescript
if (apiKey) {
  throw new Error(
    'Google Vertex transcription models do not support Express Mode API keys...',
  );
}
```

The `createSpeechModel` factory for Chirp should include the same guard so users get a clear error instead of a cryptic API failure:

```typescript
const createSpeechModel = (modelId: GoogleVertexSpeechModelId) => {
  if (modelId.startsWith('chirp')) {
    if (apiKey) {
      throw new Error(
        'Google Vertex Chirp speech models do not support Express Mode API keys. Use standard Google Cloud credentials instead.',
      );
    }
    // ... rest of Cloud TTS model creation
  }
  return new GoogleSpeechModel(modelId, createConfig('speech'));
};
```

---

### 2. Voice name edge case -- empty locale prefix -- should fix

In `google-vertex-cloud-tts-speech-model.ts:113-117`:

```typescript
languageCode =
  language ??
  voice.split(`-${CHIRP3_HD_VOICE_INFIX}`)[0] ??
  DEFAULT_LANGUAGE;
```

If someone passes a fully-qualified voice name **without** a locale prefix (e.g. `Chirp3-HD-Kore`), `split('-Chirp3-HD')[0]` returns `""` (empty string). Since `??` only checks for `null`/`undefined`, the empty string passes through as `languageCode`, which is an invalid BCP-47 locale.

Suggested fix -- use `||` instead of `??` for the split result, or add an explicit emptiness check:

```typescript
languageCode =
  language ??
  voice.split(`-${CHIRP3_HD_VOICE_INFIX}`)[0] ||
  DEFAULT_LANGUAGE;
```

---

### 3. Missing `providerMetadata` -- nice to have

The existing `GoogleSpeechModel` returns `providerMetadata` with useful information (`sampleRate`, `mimeType`). The Cloud TTS model currently returns none. Consider adding metadata for consistency and to help callers detect the audio format programmatically:

```typescript
return {
  audio,
  warnings,
  request: { body: JSON.stringify(requestBody) },
  response: { timestamp: currentDate, modelId: this.modelId, headers: responseHeaders, body: rawResponse },
  providerMetadata: {
    google: {
      mimeType: 'audio/wav',
    },
  },
};
```

---

### 4. Missing examples -- nice to have

Per contributing/building-new-features.md, new features should include usage examples in `examples/ai-functions/`. Consider adding a basic Chirp 3 HD TTS example under `examples/ai-functions/src/speech/google-vertex/`.

---

### What looks great

- The voice-name composition logic (composed vs. fully-qualified passthrough with `language` precedence) is well thought out and well-tested.
- The `WORKFLOW_SERIALIZE`/`WORKFLOW_DESERIALIZE` pattern is correctly implemented, matching the transcription model convention.
- Using `googleVertexFailedResponseHandler` is correct for a Vertex-specific model.
- The response schema uses `.nullish()` per the provider conventions.
- The changeset is properly scoped as a `patch`.
- Test coverage is thorough -- endpoint URL, headers, user-agent, request body shape, voice composition edge cases, audio decoding, speed mapping, warnings, error handling, and response metadata.

Overall, nice work! The Express Mode guard and the voice name edge case are the main items I would want addressed before merging.

**Comment by 31Carlton7** [comment]
> ## Code Review: feat(provider/google-vertex): add Chirp 3 HD voices (Cloud Text-to-Speech) speech model
> Thanks for the thorough implementation! The overall structure is solid, the tests are well-written, and the voice-name composition logic is clear. I have a few suggestions below.
> 
> ### 1. Missing Express Mode (API key) guard -- should fix
> The Cloud Text-to-Speech API (`texttospeech.googleapis.com`) is a **separate Google Cloud API**, not the Vertex AI `generateContent` endpoint. Express Mode API keys only work with the Vertex AI endpoint. The transcription model (`google-vertex-provider-base.ts:335-352`) already guards against this:
> 
> ```ts
> if (apiKey) {
>   throw new Error(
>     'Google Vertex transcription models do not support Express Mode API keys...',
>   );
> }
> ```
> 
> The `createSpeechModel` factory for Chirp should include the same guard so users get a clear error instead of a cryptic API failure:
> 
> ```ts
> const createSpeechModel = (modelId: GoogleVertexSpeechModelId) => {
>   if (modelId.startsWith('chirp')) {
>     if (apiKey) {
>       throw new Error(
>         'Google Vertex Chirp speech models do not support Express Mode API keys. Use standard Google Cloud credentials instead.',
>       );
>     }
>     // ... rest of Cloud TTS model creation
>   }
>   return new GoogleSpeechModel(modelId, createConfig('speech'));
> };
> ```
> 
> ### 2. Voice name edge case -- empty locale prefix -- should fix
> In `google-vertex-cloud-tts-speech-model.ts:113-117`:
> 
> ```ts
> languageCode =
>   language ??
>   voice.split(`-${CHIRP3_HD_VOICE_INFIX}`)[0] ??
>   DEFAULT_LANGUAGE;
> ```
> 
> If someone passes a fully-qualified voice name **without** a locale prefix (e.g. `Chirp3-HD-Kore`), `split('-Chirp3-HD')[0]` returns `""` (empty string). Since `??` only checks for `null`/`undefined`, the empty string passes through as `languageCode`, which is an invalid BCP-47 locale.
> 
> Suggested fix -- use `||` instead of `??` for the split result, or add an explicit emptiness check:
> 
> ```ts
> languageCode =
>   language ??
>   voice.split(`-${CHIRP3_HD_VOICE_INFIX}`)[0] ||
>   DEFAULT_LANGUAGE;
> ```
> 
> ### 3. Missing `providerMetadata` -- nice to have
> The existing `GoogleSpeechModel` returns `providerMetadata` with useful information (`sampleRate`, `mimeType`). The Cloud TTS model currently returns none. Consider adding metadata for consistency and to help callers detect the audio format programmatically:
> 
> ```ts
> return {
>   audio,
>   warnings,
>   request: { body: JSON.stringify(requestBody) },
>   response: { timestamp: currentDate, modelId: this.modelId, headers: responseHeaders, body: rawResponse },
>   providerMetadata: {
>     google: {
>       mimeType: 'audio/wav',
>     },
>   },
> };
> ```
> 
> ### 4. Missing examples -- nice to have
> Per contributing/building-new-features.md, new features should include usage examples in `examples/ai-functions/`. Consider adding a basic Chirp 3 HD TTS example under `examples/ai-functions/src/speech/google-vertex/`.
> 
> ### What looks great
> * The voice-name composition logic (composed vs. fully-qualified passthrough with `language` precedence) is well thought out and well-tested.
> * The `WORKFLOW_SERIALIZE`/`WORKFLOW_DESERIALIZE` pattern is correctly implemented, matching the transcription model convention.
> * Using `googleVertexFailedResponseHandler` is correct for a Vertex-specific model.
> * The response schema uses `.nullish()` per the provider conventions.
> * The changeset is properly scoped as a `patch`.
> * Test coverage is thorough -- endpoint URL, headers, user-agent, request body shape, voice composition edge cases, audio decoding, speed mapping, warnings, error handling, and response metadata.
> 
> Overall, nice work! The Express Mode guard and the voice name edge case are the main items I would want addressed before merging.

thanks for the thorough review! all four addressed now!

**Comment by R-Taneja** [comment]
this may not be valid if the change is tested and works but pasting just for visibility:

> The new Cloud TTS endpoint has no project in its URL. The current Node authentication path calls getAccessToken() and manually emits only Authorization, while the Chirp adapter also drops the provider project.
> 
> For user ADC credentials created through gcloud auth application-default login, Google commonly requires:
> 
> x-goog-user-project: PROJECT_ID
> 
> Without it, authentication can succeed while the request fails with quota-project or serviceusage.services.use errors, or quota is attributed to the wrong consumer.
> 
> The shared Google auth wrapper should probably obtain request headers through the auth client, rather than extracting only the access token, so quotaProjectId metadata is preserved. At minimum, the Cloud TTS call needs an appropriate x-goog-user-project unless the caller already supplied it.

File: .changeset/chirp-three-hd-speech.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@ai-sdk/google-vertex': patch
+---
+
+Add Chirp 3: HD voice support to the speech model via the Google Cloud Text-to-Speech API. Use `vertex.speech('chirp-3-hd')` with `voice` (e.g. `Kore`) and `language` (BCP-47 locale, e.g. `en-US`), or pass a fully-qualified voice name such as `en-US-Chirp3-HD-Kore`.
```
Comment (R-Taneja) [inline .changeset/chirp-three-hd-speech.md]: nit: convention is to match changeset descriptions to the PR title

**Review by R-Taneja — APPROVED** [review]

File: .changeset/chirp-three-hd-speech.md (hunk 1 above)
Comment (31Carlton7) [reply .changeset/chirp-three-hd-speech.md]: fixed in e59f6b4e46 — changeset now matches the PR title

**Comment by 31Carlton7** [comment]
thanks for the review! all four items addressed in f9bd6ad2e6: express-mode guard (same pattern as the transcription model), the locale edge case (the suggested `||` alone wasn't sufficient — for `Chirp3-HD-Kore` the `-Chirp3-HD` separator never matches, so split returns the whole name; reworked the extraction with a test for exactly that input), providerMetadata with `audio/wav`, and a usage example under `examples/ai-functions/src/generate-speech/google-vertex/`. 256/256 package tests green.
